OTAnalytics/plugin_datastore/track_store.py

```diff
@@ -288,22 +298,31 @@ def from_dataframe(
     ) -> "PandasTrackDataset":
         if tracks.empty:
             return PandasTrackDataset(track_geometry_factory)
+
         classified_tracks = _assign_track_classification(tracks, calculator)
+        newly_indexed_tracks = classified_tracks.set_index(
+            track.TRACK_CLASSIFICATION, append=True
+        ).reorder_levels([LEVEL_CLASSIFICATION, LEVEL_TRACK_ID, LEVEL_OCCURRENCE])
+
         return PandasTrackDataset(
             track_geometry_factory,
-            classified_tracks,
+            newly_indexed_tracks,
             geometry_datasets=geometry_dataset,
         )
 
     def add_all(self, other: Iterable[Track]) -> "PandasTrackDataset":
         new_tracks = self.__get_tracks(other)
         if new_tracks.empty:
             return self
+        if LEVEL_CLASSIFICATION in new_tracks.index.names:
+            new_tracks = new_tracks.reset_index(level=LEVEL_CLASSIFICATION, drop=True)
         if self._dataset.empty:
             return PandasTrackDataset.from_dataframe(
                 new_tracks, self.track_geometry_factory, calculator=self.calculator
             )
-        updated_dataset = pandas.concat([self._dataset, new_tracks]).sort_index()
+        updated_dataset = pandas.concat(
+            [self._dataset.reset_index(LEVEL_CLASSIFICATION, drop=True), new_tracks]
```

Comment:
The problem is that by concatenating two DataFrames with differing multi-index levels will mess up our index. Our index becomes a regular single level index containing the previous multi-index values as tuples of different sizes.  

 It is not enforced that `new_tracks` has `LEVEL_CLASSIFCATION` as an index level. To enforce that the max classification must be calculated for `new_tracks`, which would be unnecessary since the max classification will be calculated again when creating a new instance of `PandasTrackDataset` with `from_dataframe`.